src/rust/ide/ast/impl/src/crumbs.rs

```diff
@@ -0,0 +1,562 @@
+        let new_ast = self.ast().set(crumb,new_ast)?;
+        let ret = known::KnownAst::try_new(new_ast)?;
+        Ok(ret)
+    }
+
+    fn iter_subcrumbs<'a>(&'a self) -> Box<dyn Iterator<Item = Self::Crumb> + 'a> {
+        self.ast().iter_subcrumbs()
+    }
+}
+
+
+
+// ===========================
+// === Recursive Traversal ===
+// ===========================
+
+/// Interface for recursive AST traversal using `Crumb` sequence.
+///
+/// Intended for `Ast` and `Ast`-like types, like `KnownAst`.
+pub trait TraversableAst {
+    /// Returns rewritten AST where child AST under location designated by `crumbs` is updated.
+    ///
+    /// Works recursively.
+    fn set_traversing(&self, crumbs:&[Crumb], new_ast:Ast) -> FallibleResult<Self>
+    where Self:Sized {
+        let ast         = self.ast_ref();
+        let updated_ast = if let Some(first_crumb) = crumbs.first() {
+            let child = ast.get(first_crumb)?;
+            let updated_child = child.set_traversing(&crumbs[1..], new_ast)?;
+            ast.set(first_crumb,updated_child)
+        } else {
+            Ok(new_ast)
+        };
+        Self::from_ast(updated_ast?)
+    }
+
+    /// Recursively traverses AST to retrieve AST node located by given crumbs sequence.
+    fn get_traversing<'a>(&'a self, crumbs:&[Crumb]) -> FallibleResult<&'a Ast> {
+        let ast = self.ast_ref();
+        if let Some(first_crumb) = crumbs.first() {
+            let child = ast.get(first_crumb)?;
+            child.get_traversing(&crumbs[1..])
+        } else {
+            Ok(ast)
+        }
+    }
+
+    /// Access this node's AST.
+    fn ast_ref(&self) -> &Ast;
+
+    /// Wrap Ast into Self.
+    fn from_ast(ast:Ast) -> FallibleResult<Self> where Self:Sized;
+}
+
+impl TraversableAst for Ast {
+    fn ast_ref(&self) -> &Ast { self }
+
+    fn from_ast(ast:Ast) -> FallibleResult<Self> { Ok(ast) }
+}
+
+impl<T,E> TraversableAst for known::KnownAst<T>
+where for<'t> &'t Shape<Ast> : TryInto<&'t T, Error=E>,
+      E                      : failure::Fail {
+    fn ast_ref(&self) -> &Ast { self.ast() }
+
+    fn from_ast(ast:Ast) -> FallibleResult<Self> { Ok(ast.try_into()?) }
+}
+
+
+
+// ===============
+// === Utility ===
+// ===============
+
+/// Iterates over indices of non-empty lines in a line sequence.
+pub fn non_empty_line_indices<'a, T:'a>
+(iter:impl Iterator<Item = &'a crate::BlockLine<Option<T>>> + 'a)
+ -> impl Iterator<Item=usize> + 'a {
+    iter.enumerate().filter_map(|(line_index,line)| {
+        line.elem.as_ref().map(|_| line_index)
+    })
+}
+
+
+
+// ===============
+// === Located ===
+// ===============
+
+/// Item which location is identified by `Crumbs`.
+#[derive(Clone,Debug,Shrinkwrap)]
+pub struct Located<T> {
+    /// Crumbs from containing parent.
+    pub crumbs : Crumbs,
+    /// The sub-item representation.
+    #[shrinkwrap(main_field)]
+    pub item   : T
+}
+
+impl<T> Located<T> {
```

Comment:
I'm feeling uneasy about this structure, but I will explain it during the call.

Reply:
Explained and finally accepted to leave as-is, despite lingering uneasiness.